### PR TITLE
Improve control over output formatting

### PR DESCRIPTION
**Pull request by michaelquinn32**
* Store a column of formatted values in the skim_df
* Add an interface for setting formats for types of columns
* Fix tests for new column
* Update print methods to work with formatted column

### CODE REVIEW

File: R/options.R [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+#' For skimming functions, we show just their names. The function bodies
+#' can be confusing. For formatting options, we show everything.
+#' 
+#' @param which A character vector. One or more of the classes whose summary
+#'  functions/ formatting options you wish to display.
+#' @return A list. The names of the list match the classes that have assigned
+#'  summary functions or formatting options.
+#' @noRd
+
+show_options <- function(which = NULL, env) {
```
Comment (elinw) [inline R/options.R]: This is really great.

File: R/stats.R [hunk 2]
```diff
@@ -22,23 +22,34 @@ n_complete <- function(x) {
...
+#' Create a contingency table and arrange its levels in descending order
+#' 
+#' @param x An object that can be interpreted as a factor (including logical)
+#' @return A "table" object, which will be treated as a named numeric vector
+#' @export
+
+sorted_count <- function(x) {
+  tab <- table(x, useNA = "always")
+  out <- setNames(as.integer(tab), names(tab))
+  sort(out, decreasing = TRUE)
```
Comment (elinw) [inline R/stats.R]: You'd need separate treatment for ordered factors that doesn't do this sorting.

File: R/stats.R [hunk 3]
```diff
@@ -22,23 +22,34 @@ n_complete <- function(x) {
...
-  out <- 0
-  if ( !all(x == 0) & length(x) != 0) {
-    hist_dt <- table(cut(x, 10))
-    hist_dt <- hist_dt / max(hist_dt)
-    names(out) <- colformat::spark_bar(hist_dt)
-    return(out)
-  }
-  names(out) <- ""
-  return(out)
+  # Handle empty vectors
```
Comment (elinw) [inline R/stats.R]: I believe the NA handling was because at the unconf we found a big problem with all NA columns and this function.

Comment (elinw) [reply R/stats.R]: Not handled anywhere at this time but we probably want to have a maximum length here as we do with the line graph. Such as 40 characters or something like that.

File: R/stats.R (hunk 2 above)
Comment (michaelquinn32) [reply R/stats.R]: We're sorting the counts, which is separate from the ordering of the factor itself.

Imagine that we're summarizing a table with a factor with 100 levels. The previous could function would try and show counts for every level. That's not helpful, and it goes against the "skim" concept of the package.

Now, we just take the most popular levels and show them in the results. The formatting options .levels controls how many these to show.

File: R/stats.R (hunk 3 above)
Comment (michaelquinn32) [reply R/stats.R]: A later commit changes this to use the character formatting options within the package.

File: R/stats.R (hunk 2 above)
Comment (elinw) [reply R/stats.R]: I agree about limiting the number of categories in both ordered and regular factors, that's why in the branch they are limited to 45 characters.  But ordering in ordered factors has real meaning, I don't think it makes sense to say small, large, medium, xsmall, xxl.   I think right now we are not doing anything at all to indicated that a factor is ordered or not either so in terms of skimming that's actually a problem to consider if we are going to throw them in with the regular factors. That's information you would want to know in the data set.  Maybe we can add something when creating this string?

File: R/stats.R (hunk 3 above)
Comment (elinw) [reply R/stats.R]: Let's make sure to have a test for the all NA case (really we should for all of the types).

File: R/functions.R [hunk 4]
```diff
@@ -100,55 +62,166 @@ skim_with_defaults <- function() {
...
+  complete = n_complete,
+  n = length,
+  mean = purrr::partial(mean, na.rm = TRUE),
+  sd = purrr::partial(sd, na.rm = TRUE),
+  min = purrr::partial(min, na.rm = TRUE),
+  p25 = purrr::partial(quantile, probs = .25, na.rm = TRUE, names = FALSE),
+  median = purrr::partial(median, na.rm = TRUE),
+  p75 = purrr::partial(quantile, probs = .75, na.rm = TRUE, names = FALSE),
+  max = purrr::partial(max, na.rm = TRUE),
+  hist = inline_hist
```
Comment (elinw) [inline R/functions.R]: What are the performance implications of going to quantile twice as opposed to once to get all three (or even all 5)?

Comment (michaelquinn32) [reply R/functions.R]: Haven't thought about it TBH. My bigger concern was solving the issues with output.

If a function produces a single value, skim adds it as a column. If a functions produces multiple values, skim now produces name: value pairs. I think this looks really nice for factors, which don't have a high impact graphic like the histogram.

**Comment by michaelquinn32** [comment]
This gives us grouped data frames. The change looks pretty big, but the code for grouped frames is actually pretty simple. Doing it this way means that we don't even have to support a separate class. Skim will be able to handle printing `skim_df` objects regardless of whether the original df was grouped.

**Comment by michaelquinn32** [comment]
Here's the current output for the ungrouped version:

``` r
library(skimr)
skim(iris)
#> Skim summary statistics
#>  n obs: 52 
#>  n variables: 6 
#> 
#> Variable type: factor 
#>       var missing complete   n n_unique                       top_counts
#> 1 Species       0      150 150        3 set: 50, ver: 50, vir: 50, NA: 0
#> 
#> Variable type: numeric 
#>            var missing complete   n mean   sd  min  p25 median  p75  max
#> 1 Petal.Length       0      150 150 3.76 1.77 1.00 1.60   4.35 5.10 6.90
#> 2  Petal.Width       0      150 150 1.20 0.76 0.10 0.30   1.30 1.80 2.50
#> 3 Sepal.Length       0      150 150 5.84 0.83 4.30 5.10   5.80 6.40 7.90
#> 4  Sepal.Width       0      150 150 3.06 0.44 2.00 2.80   3.00 3.30 4.40
#>         hist
#> 1 ▇▃▁▁▂▆▆▃▂▁
#> 2 ▇▂▁▂▅▃▁▅▂▃
#> 3 ▂▇▅▇▆▆▅▂▂▂
#> 4 ▁▁▃▃▇▃▂▂▁▁
```

**Comment by michaelquinn32** [comment]
And the grouped version:

``` r
library(skimr)
skim(dplyr::group_by(iris, Species))
#> Skim summary statistics
#>  n obs: 156 
#>  n variables: 7 
#>  group variables: Species 
#> 
#> Variable type: factor 
#>      Species     var missing complete  n n_unique
#> 1     setosa Species       0       50 50        1
#> 2 versicolor Species       0       50 50        1
#> 3  virginica Species       0       50 50        1
#>                       top_counts
#> 1 set: 50, ver: 0, vir: 0, NA: 0
#> 2 ver: 50, set: 0, vir: 0, NA: 0
#> 3 vir: 50, set: 0, ver: 0, NA: 0
#> 
#> Variable type: numeric 
#>       Species          var missing complete  n mean   sd  min  p25 median
#> 1      setosa Petal.Length       0       50 50 1.46 0.17 1.00 1.40   1.50
#> 2      setosa  Petal.Width       0       50 50 0.25 0.11 0.10 0.20   0.20
#> 3      setosa Sepal.Length       0       50 50 5.01 0.35 4.30 4.80   5.00
#> 4      setosa  Sepal.Width       0       50 50 3.43 0.38 2.30 3.20   3.40
#> 5  versicolor Petal.Length       0       50 50 4.26 0.47 3.00 4.00   4.35
#> 6  versicolor  Petal.Width       0       50 50 1.33 0.20 1.00 1.20   1.30
#> 7  versicolor Sepal.Length       0       50 50 5.94 0.52 4.90 5.60   5.90
#> 8  versicolor  Sepal.Width       0       50 50 2.77 0.31 2.00 2.52   2.80
#> 9   virginica Petal.Length       0       50 50 5.55 0.55 4.50 5.10   5.55
#> 10  virginica  Petal.Width       0       50 50 2.03 0.27 1.40 1.80   2.00
#> 11  virginica Sepal.Length       0       50 50 6.59 0.64 4.90 6.23   6.50
#> 12  virginica  Sepal.Width       0       50 50 2.97 0.32 2.20 2.80   3.00
#>     p75  max       hist
#> 1  1.58 1.90 ▁▁▁▅▇▇▅▂▁▁
#> 2  0.30 0.60 ▂▇▁▂▁▂▁▁▁▁
#> 3  5.20 5.80 ▂▃▁▃▇▅▂▅▁▂
#> 4  3.68 4.40 ▁▁▁▅▃▇▃▃▁▁
#> 5  4.60 5.10 ▁▂▂▂▆▆▅▇▆▃
#> 6  1.50 1.80 ▅▂▃▇▅▁▆▂▁▁
#> 7  6.30 7.00 ▃▁▅▇▃▆▃▂▃▂
#> 8  3.00 3.40 ▁▂▅▃▆▅▅▇▂▂
#> 9  5.88 6.90 ▁▃▇▂▇▃▃▁▁▂
#> 10 2.30 2.50 ▂▁▁▇▃▅▅▂▆▅
#> 11 6.90 7.90 ▁▁▃▃▇▆▃▃▁▃
#> 12 3.18 3.80 ▁▂▁▇▇▂▅▁▁▁
```

**Comment by michaelquinn32** [comment]
Here's an updated output.

``` r
skimr::skim(iris)
#> Skim summary statistics
#>  n obs: 52 
#>  n variables: 6 
#> 
#> Variable type: factor 
#>       var missing complete   n n_unique                       top_counts
#> 1 Species       0      150 150        3 set: 50, ver: 50, vir: 50, NA: 0
#> 
#> Variable type: numeric 
#>            var missing complete   n mean   sd min p25 median p75 max
#> 1 Petal.Length       0      150 150 3.76 1.77 1   1.6   4.35 5.1 6.9
#> 2  Petal.Width       0      150 150 1.2  0.76 0.1 0.3   1.3  1.8 2.5
#> 3 Sepal.Length       0      150 150 5.84 0.83 4.3 5.1   5.8  6.4 7.9
#> 4  Sepal.Width       0      150 150 3.06 0.44 2   2.8   3    3.3 4.4
#>       hist
#> 1 ▇▁▁▂▅▅▃▁
#> 2 ▇▁▁▅▃▃▂▂
#> 3 ▂▇▅▇▆▅▂▂
#> 4 ▁▂▅▇▃▂▁▁
```

**Comment by michaelquinn32** [comment]
mtcars also looks pretty nice

``` r
skimr::skim(mtcars)
#> Skim summary statistics
#>  n obs: 121 
#>  n variables: 6 
#> 
#> Variable type: numeric 
#>     var missing complete  n   mean     sd   min    p25 median    p75
#> 1    am       0       32 32   0.41   0.5   0      0      0      1   
#> 2  carb       0       32 32   2.81   1.62  1      2      2      4   
#> 3   cyl       0       32 32   6.19   1.79  4      4      6      8   
#> 4  disp       0       32 32 230.72 123.94 71.1  120.83 196.3  326   
#> 5  drat       0       32 32   3.6    0.53  2.76   3.08   3.7    3.92
#> 6  gear       0       32 32   3.69   0.74  3      3      4      4   
#> 7    hp       0       32 32 146.69  68.56 52     96.5  123    180   
#> 8   mpg       0       32 32  20.09   6.03 10.4   15.43  19.2   22.8 
#> 9  qsec       0       32 32  17.85   1.79 14.5   16.89  17.71  18.9 
#> 10   vs       0       32 32   0.44   0.5   0      0      0      1   
#> 11   wt       0       32 32   3.22   0.98  1.51   2.58   3.33   3.61
#>       max     hist
#> 1    1    ▇▁▁▁▁▁▁▆
#> 2    8    ▆▇▂▇▁▁▁▁
#> 3    8    ▆▁▁▃▁▁▁▇
#> 4  472    ▇▆▁▂▅▃▁▂
#> 5    4.93 ▃▇▁▅▇▂▁▁
#> 6    5    ▇▁▁▆▁▁▁▂
#> 7  335    ▃▇▃▅▂▃▁▁
#> 8   33.9  ▃▇▇▇▃▂▂▂
#> 9   22.9  ▃▂▇▆▃▃▁▁
#> 10   1    ▇▁▁▁▁▁▁▆
#> 11   5.42 ▃▃▃▇▆▁▁▂
```

File: R/functions.R (hunk 4 above)
Comment (elinw) [reply R/functions.R]: That's the actual issue with median and why I was thinking that we switch to calling it 50th percentile.  Also I want to deal with time series and that (but that's a small issue).

**Comment by elinw** [comment]
I think this looks great.  I'm going to pull down the branch and run through some of my local examples but I really love the handling of the formatting.

I was wondering about the question of having a separate class for grouped which was @GShotwell 's original concept.  I think if we can avoid it that solve a separate complexity issue because if someone makes their own skim object (e.g. for simple features like in the vignette) and then that has grouping, what happens.    

I think we still also need to think about ordered factors and whether we need to at least mark them as ordered.

**Comment by elinw** [comment]
@michaelquinn32 

I think we should exclude any grouping variable from rows of the skim objects.
They aren't really variables within the groups.

**Comment by elinw** [comment]
Also the number of rows and the number of variables are not correct ... there are only 5 variables in iris and there are 150 observations.

**Comment by michaelquinn32** [comment]
After merging in all  of Elin's excellent changes, here's where we stand.

This is the standard version:

``` r
library(skimr)
skim(iris)
#> Skim summary statistics
#>  n obs: 150 
#>  n variables: 5 
#> 
#> Variable type: factor 
#>       var missing complete   n n_unique                       top_counts
#> 1 Species       0      150 150        3 set: 50, ver: 50, vir: 50, NA: 0
#>   ordered
#> 1   FALSE
#> 
#> Variable type: numeric 
#>            var missing complete   n mean   sd min p25 median p75 max
#> 1 Petal.Length       0      150 150 3.76 1.77 1   1.6   4.35 5.1 6.9
#> 2  Petal.Width       0      150 150 1.2  0.76 0.1 0.3   1.3  1.8 2.5
#> 3 Sepal.Length       0      150 150 5.84 0.83 4.3 5.1   5.8  6.4 7.9
#> 4  Sepal.Width       0      150 150 3.06 0.44 2   2.8   3    3.3 4.4
#>       hist
#> 1 ▇▁▁▂▅▅▃▁
#> 2 ▇▁▁▅▃▃▂▂
#> 3 ▂▇▅▇▆▅▂▂
#> 4 ▁▂▅▇▃▂▁▁
```

This is the grouped version:

``` r
library(skimr)
dplyr::group_by(iris, Species) %>% skim()
#> Skim summary statistics
#>  n obs: 
#>  n variables: 
#>  group variables: Species 
#> 
#> Variable type: numeric 
#>       Species          var missing complete  n mean   sd min  p25 median
#> 1      setosa Petal.Length       0       50 50 1.46 0.17 1   1.4    1.5 
#> 2      setosa  Petal.Width       0       50 50 0.25 0.11 0.1 0.2    0.2 
#> 3      setosa Sepal.Length       0       50 50 5.01 0.35 4.3 4.8    5   
#> 4      setosa  Sepal.Width       0       50 50 3.43 0.38 2.3 3.2    3.4 
#> 5  versicolor Petal.Length       0       50 50 4.26 0.47 3   4      4.35
#> 6  versicolor  Petal.Width       0       50 50 1.33 0.2  1   1.2    1.3 
#> 7  versicolor Sepal.Length       0       50 50 5.94 0.52 4.9 5.6    5.9 
#> 8  versicolor  Sepal.Width       0       50 50 2.77 0.31 2   2.52   2.8 
#> 9   virginica Petal.Length       0       50 50 5.55 0.55 4.5 5.1    5.55
#> 10  virginica  Petal.Width       0       50 50 2.03 0.27 1.4 1.8    2   
#> 11  virginica Sepal.Length       0       50 50 6.59 0.64 4.9 6.23   6.5 
#> 12  virginica  Sepal.Width       0       50 50 2.97 0.32 2.2 2.8    3   
#>     p75 max     hist
#> 1  1.58 1.9 ▁▁▅▇▇▅▂▁
#> 2  0.3  0.6 ▂▇▁▂▂▁▁▁
#> 3  5.2  5.8 ▂▃▅▇▇▃▁▂
#> 4  3.68 4.4 ▁▁▃▅▇▃▂▁
#> 5  4.6  5.1 ▁▃▂▆▆▇▇▃
#> 6  1.5  1.8 ▆▃▇▅▆▂▁▁
#> 7  6.3  7   ▃▂▇▇▇▃▅▂
#> 8  3    3.4 ▁▂▃▅▃▇▃▁
#> 9  5.88 6.9 ▂▇▃▇▅▂▁▂
#> 10 2.3  2.5 ▂▁▇▃▃▆▅▃
#> 11 6.9  7.9 ▁▁▃▇▅▃▂▃
#> 12 3.18 3.8 ▁▃▇▇▅▃▁▂
```

**Comment by elinw** [comment]
Okay, I'm going to push the merge button.

**Pull request merged**